api/core/workflow/nodes/knowledge_retrieval/knowledge_retrieval_node.py

```diff
@@ -73,30 +75,31 @@ def _run(self, variable_pool: VariablePool) -> NodeRunResult:
 
     def _fetch_dataset_retriever(self, node_data: KnowledgeRetrievalNodeData, query: str) -> list[
         dict[str, Any]]:
-        """
-        A dataset tool is a tool that can be used to retrieve information from a dataset
-        :param node_data: node data
-        :param query: query
-        """
-        tools = []
         available_datasets = []
         dataset_ids = node_data.dataset_ids
-        for dataset_id in dataset_ids:
-            # get dataset from dataset id
-            dataset = db.session.query(Dataset).filter(
-                Dataset.tenant_id == self.tenant_id,
-                Dataset.id == dataset_id
-            ).first()
 
-            # pass if dataset is not available
-            if not dataset:
-                continue
+        # Subquery: Count the number of available documents for each dataset
+        subquery = db.session.query(
+            Document.dataset_id,
+            func.count(Document.id).label('available_document_count')
+        ).filter(
+            Document.indexing_status == 'completed',
+            Document.enabled == True,
+            Document.archived == False
+        ).group_by(Document.dataset_id).subquery()
+
+        results = db.session.query(Dataset).join(
+            subquery, Dataset.id == subquery.c.dataset_id
+        ).filter(
+            Dataset.tenant_id == self.tenant_id,
+            Dataset.id.in_(dataset_ids),
+            subquery.c.available_document_count > 0,
```

Comment:
How about moving the filter condition of document count >0 into the subquery, making the subquery return only the dataset with at least one doc, this could be a better way to push down for planning and fetching less mid-result.

Reply:
Good idea, I made some optimizations.

Reply:
Great.